### ISSUE TITLE
ngx.timer.at memory leak concern

### ISSUE THREAD

**Issue by atrifan** [openresty/openresty#373]
We noticed that ngx.timer.at produces a memory increase that never gets deallocated. The behaviour is seen even better after nginx reload.

Do you know of any problems with the ngx.timer.at function. Is this expected? How can this be avoided?

We have tested with nginx 1.9.7.3 and 1.13.6.1. We tested on both docker - alpine 3.4 and macos sierra. The scenario is very easy to reproduce. I am posting the details bellow with a testing configuration and some other information

Nginx versions: 1.9.7.3 / 1.13.6.1

OSs: alpine 3.4 (Docker) / macos sierra

Basic configuration to reproduce scenario:
```
user                 root;
worker_processes     32;
worker_rlimit_nofile 524288;

events {
	use epoll;
	worker_connections 5000;
	multi_accept on;
}

http {
    init_worker_by_lua_block {
        x = function(premature)
            if premature then
                return
            end
            ngx.timer.at(0.4, x)
        end
        ngx.timer.at(0.4, x)
    }
}
```

Openresty compile parameters:

```
 ./configure \
            --with-stream \
            --with-stream_ssl_module \
            --with-http_ssl_module \
            --with-http_stub_status_module \
            --with-http_realip_module \
            --with-http_addition_module \
            --with-http_sub_module \
            --with-http_dav_module \
            --with-http_gunzip_module  \
            --with-http_gzip_static_module \
            --with-http_auth_request_module \
            --with-http_random_index_module \
            --with-http_secure_link_module \
            --with-http_degradation_module \
            --with-http_auth_request_module  \
            --with-http_v2_module \
            --with-luajit \
            --without-http_ssi_module \
            --without-http_userid_module \
            --without-http_uwsgi_module \
            -with-http_geoip_module \
            --without-http_scgi_module \
            --without-http_rewrite_module \
            --with-openssl=... \
            --with-pcre=...\
            --with-pcre-jit=..\
            -j6 
```

**Comment by agentzh** [comment]
@atrifan I cannot reproduce any leaks using your example above. I even decreased the timer delay from 0.4 to 0.001, still seeing nothing. And I also verified from the nginx error log file that the lua timer is indeed running. I use the following command to observe the (only) nginx worker process's memory usage over time:

```
top -p `pgrep -f 'nginx: worker'`
```

And all the memory metrics for that nginx worker process never changes over time:

```
   PID USER      PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND
 42616 agentzh   20   0   68164   5252   4048 S   6.7  0.1   0:01.15 nginx

   PID USER      PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND
 42616 agentzh   20   0   68164   5252   4048 S   5.9  0.1   0:18.61 nginx
```

I'm using the latest OpenResty 1.13.6.2 release BTW.

**Comment by agentzh** [comment]
@atrifan I'm not using any docker stuff on my side though. Maybe it's a problem specific to docker or even a docker bug?

**Comment by atrifan** [comment]
Posted two screen shots from 1.9.7.3 with the issue reproduced. 
The same issue reproduces in 1.13.6.1

<img width="584" alt="screen shot 2018-05-30 at 10 16 59" src="https://user-images.githubusercontent.com/2162576/40716523-f70b50ac-6411-11e8-86f1-b7adddc536ca.png">
<img width="579" alt="screen shot 2018-05-30 at 10 18 32" src="https://user-images.githubusercontent.com/2162576/40716524-f727fdb0-6411-11e8-8a8c-0d19108a719e.png">'

Checked 1.13.6.2 - the issue does not reproduce anymore. 

You can mark this as closed with fixed in 1.13.6.2.

Thanks

**Comment by agentzh** [comment]
@atrifan You can close this issue yourself BTW.

This is also why the user should always try the latest version before reporting an issue to avoid wasting time on both sides ;)
